### ISSUE TITLE
Tuple rest descriptor not considered in array lang lib functions

### ISSUE THREAD

**Issue by pubudu91** [ballerina-platform/ballerina-lang#18660]
Consider the following code snippet:
```ballerina
function testPushInTuples() returns [int, string, float...] {
    [int, string, float...] tup = [10, "Ballerina"];
    tup.push(12.34, 45.67);
    return tup;
}
```
For the above, the following compile errors are given
```
ERROR: .::arraylib_test.bal:320:5:: incompatible types: expected '(int|string)', found 'float'
ERROR: .::arraylib_test.bal:320:5:: incompatible types: expected '(int|string)', found 'float'
```
In the above case, the push() function should only be expecting `float` values.

**Comment by pubudu91** [comment]
Related issue #18662

**Comment by MaryamZi** [comment]
Another scenarios

```ballerina
import ballerina/io;

public function main() {
    [string, int...] x = ["hello", 1, 2];

    [int, string][] y = x.enumerate(); // Should be [int, string|int][] but doesn't fail now

    foreach [int, string] [i, v] in y {
        io:println("index: ", i, ", value: ", v); // results in a TypeCastError at index 1
    }
}
```

**Comment by hasithaa** [comment]
Fixed with #19348
